src/main/java/backend/RepoIO.java

```diff
@@ -51,14 +52,32 @@ public RepoIO(boolean isTestMode, boolean enableTestJSON) {
 
     public CompletableFuture<Model> openRepository(String repoId) {
         if (repoStore.isRepoStored(repoId)) {
-            return repoStore.loadRepository(repoId)
-                .thenCompose(this::updateModel)
-                .exceptionally(withResult(new Model(repoId)));
+            return loadRepositoryFromStore(repoId);
         } else {
-            return repoSource.downloadRepository(repoId)
+            return downloadRepositoryFromSource(repoId);
+        }
+    }
+
+    private CompletableFuture<Model> loadRepositoryFromStore(String repoId) {
+        return repoStore.loadRepository(repoId)
+                .thenCompose(this::updateModel)
+                .exceptionally(ex -> {
+                    try {
```

Comment:
Yup, I think that' the case. You can confirm it *[here](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletionStage.html)*, it's the 4th paragraph of the intro description) but I may misinterpret something. What I was trying actually has a slightly difference purpose though. I wanted to use `thenCompose(downloadRepo...) with `exceptionally` and some additional stages before and after to handle the downloading so we can  eliminate the `try/catch` and blocking `get`` all together, but then that `thenCompose` complete exceptionally if 1 of the stage before complete exceptionally so it ends up in the last `exceptionally` and return empty model anyway